### ISSUE TITLE
Select query returns mangled data.

### ISSUE THREAD

**Issue by EelhandsMcCoy** [alexwiese/EntityFrameworkCore.OpenEdge#12]
Hi. I scaffolded with version 1.08 and had the same "OpenEdge does not support bool mapping" problem so I fixed the issue myself and then re-scaffolded.

I began to develop an API and came across this issue where I call a certain table to list.

This code:

`var r = _context.PromoTransHdr.Where(hdr => hdr.ContractNo == "Jeff").ToList();`

Results in this query:

`SELECT "hdr"."contract-no", "hdr"."expire-date", "hdr"."Billing-code", "hdr"."description", "hdr"."orig-amt", "hdr"."orig_date", "hdr"."promo-id", "hdr"."promo-transno", "hdr"."restricted", "hdr"."trans-date", "hdr"."used-amt"
FROM "pub"."promo-trans-hdr" AS "hdr"
WHERE "hdr"."contract-no" = 'Jeff'`

If I run this query in Progress' developer studio's db navigator, I get 269 rows of data, all correct, unique data. However, in my API, I get 269 results containing repeated rows and certain rows are omitted completely.

I tried recreating a new project, this time net core 2.1/asp net core and added only your package, EntityFrameworkCore.OpenEdge, the new pre-release 1.09rc1, and rescaffolded. The same issue exists, though I am able to scaffold without modification.

**Comment by alexwiese** [comment]
Hi @EelhandsMcCoy 

Check out https://github.com/alexwiese/EntityFrameworkCore.OpenEdge#gotchas

Basically EF Core thinks they are all the same record as the "ID" is the same.
You need a unique column to act as the ID. If you don't have one then add RowId as per that link

**Comment by EelhandsMcCoy** [comment]
Huh. I read that, but these tables have "keys" in that they have an index...but I suppose it's not actually unique. I'll add that to each of my models and see what happens, thank you.

**Comment by EelhandsMcCoy** [comment]
So I added:
        [Key]
        [Column("rowid")]
        public string Rowid { get; set; }

to the model class but I did not modify the context to include this column. This is the modelbuilder's entity definition:

```
            modelBuilder.Entity<PromoTransHdr>(entity =>
            {
                entity.HasKey(e => new { e.ContractNo, e.ExpireDate });

                entity.ToTable("promo-trans-hdr");

                entity.Property(e => e.ContractNo)
                    .HasColumnName("contract-no")
                    .HasColumnType("character");

                entity.Property(e => e.ExpireDate)
                    .HasColumnName("expire-date")
                    .HasColumnType("date");

                entity.Property(e => e.BillingCode)
                    .HasColumnName("Billing-code")
                    .HasColumnType("character");

                entity.Property(e => e.Description)
                    .IsRequired()
                    .HasColumnName("description")
                    .HasColumnType("character");

                entity.Property(e => e.OrigAmt)
                    .HasColumnName("orig-amt")
                    .HasDefaultValueSql("0");

                entity.Property(e => e.OrigDate)
                    .HasColumnName("orig_date")
                    .HasColumnType("date");

                entity.Property(e => e.PromoId)
                    .HasColumnName("promo-id")
                    .HasDefaultValueSql("0");

                entity.Property(e => e.PromoTransno)
                    .HasColumnName("promo-transno")
                    .HasDefaultValueSql("0");

                entity.Property(e => e.Restricted)
                    .HasColumnName("restricted")
                    .HasColumnType("logical")
                    .HasDefaultValueSql("no");

                entity.Property(e => e.TransDate)
                    .HasColumnName("trans-date")
                    .HasColumnType("date");

                entity.Property(e => e.UsedAmt)
                    .HasColumnName("used-amt")
                    .HasDefaultValueSql("0");
            });
```

This did not solve the problem.

```
{
  "value": [
    {
      "rowid": "0000000000105448618",
      "promoId": 103,
      "transDate": "2015-11-30T00:00:00",
      "description": "WarrantyPayment:1170389",
      "contractNo": "Jeff",
      "expireDate": "2015-12-30T00:00:00",
      "origAmt": 6,
      "usedAmt": 6,
      "promoTransno": 137972,
      "billingCode": "GBB",
      "restricted": false,
      "origDate": null
    },
    {
      "rowid": "0000000000105448618",
      "promoId": 103,
      "transDate": "2015-11-30T00:00:00",
      "description": "WarrantyPayment:1170389",
      "contractNo": "Jeff",
      "expireDate": "2015-12-30T00:00:00",
      "origAmt": 6,
      "usedAmt": 6,
      "promoTransno": 137972,
      "billingCode": "GBB",
      "restricted": false,
      "origDate": null
    },
// Repeat dozens of these.
}
```

The "key" that EF has selected (ContractNo, ExpireDate) is actually pretty nonsensical. The key should be PromoTransno. That is also an index in the database, but it is the second index and is not marked as 'primary.'

I will try adding rowId to the modelBuilder and if that doesn't work I'll see if I can use the second index instead.

**Comment by EelhandsMcCoy** [comment]
Adding:
```
                entity.HasKey(e => e.RowId);

                entity.Property(e => e.RowId)
                    .HasColumnName("rowid");
```

And commenting out the bad primary index/"key" fixed the problem. Thanks a lot for your help!

**Comment by alexwiese** [comment]
Great to hear that solved your problem. Yeah if there is no unique, primary index defined then EF really has no way of know which index it should be using. Best bet is to always use rowid to be sure. But make sure you’re sending rowid to/from the API to ensure updates/deletes happen correctly. Otherwise change the DB to give it a unique primary index

On 6 Sep 2019, at 11:53 pm, EelhandsMcCoy <notifications@github.com<mailto:notifications@github.com>> wrote:


Closed #12<https://github.com/alexwiese/EntityFrameworkCore.OpenEdge/issues/12>.
